File: go/881.go
import "sort"

func numRescueBoats(people []int, limit int) int {
	n := len(people)
	sort.Ints(people)
	i, j := 0, n-1
	boats := 0
	for i < n && j >= 0 && i <= j {
		if people[i]+people[j] <= limit {
			i++
		}
		j--
		boats++
	}
	return boats
}

File: python/881.py
class Solution:
    def numRescueBoats(self, people: List[int], limit: int) -> int:
        people.sort(reverse = True)
        lo,hi = 0,len(people)-1
        # boats = 0
        while lo <= hi:
            # if lo == hi:
            #     boats += 1
            if people[lo] >= limit:
                lo += 1
                # boats += 1
                continue
            else:
                if people[lo] + people[hi] > limit:
                    lo += 1
                    # boats += 1
                    continue
                else:
                    lo += 1
                    # boats += 1
                    hi -= 1
        return lo